libretto_openai/completions_impl.py

```diff
@@ -0,0 +1,252 @@
+
+        if libretto["prompt_template_name"] is None and not self.config.allow_unnamed_prompts:
+            return self._original_create(*original_args, **original_kwargs)
+
+        model_params = self._build_model_params(**original_kwargs)
+
+        # Redact PII from template parameters if configured to do so
+        self._redact_template_params(libretto["template_params"])
+
+        with event_session(
+            project_key=libretto["project_key"],
+            api_key=libretto["api_key"],
+            prompt_template_name=libretto["prompt_template_name"],
+            model_params=model_params,
+            prompt_template_text=libretto["template_text"],
+            prompt_template_chat=libretto["template_chat"],
+            chat_id=libretto["chat_id"],
+            prompt_template_params=libretto["template_params"],
+            prompt_event_id=libretto["event_id"],
+            parent_event_id=libretto["parent_event_id"],
+            feedback_key=libretto["feedback_key"],
+        ) as complete_event:
+            response = self._original_create(*original_args, **original_kwargs)
+            return_response, event_response = self._get_result(response)
+
+            # Can only do this for non-streamed responses right now
+            if hasattr(return_response, "model_copy"):
+                return_response = return_response.model_copy(
+                    update={"libretto_feedback_key": libretto["feedback_key"]}
+                )
+
+            # Redact PII before recording the event
+            event_response = self._redact_response(event_response)
+
+            complete_event(event_response)
+
+            return return_response
+
+    def _prepare_create_params(
+        self,
+        libretto: LibrettoCreateParamDict | None = None,
+        **_original_kwargs,
+    ) -> LibrettoCreateParamDict:
+        if libretto is None:
+            libretto = LibrettoCreateParams()
+        else:
+            # Don't mutate the input dict
+            libretto = libretto.copy()
+
+        libretto["prompt_template_name"] = (
+            libretto["prompt_template_name"]
+            or self.config.prompt_template_name
+            or os.environ.get("LIBRETTO_TEMPLATE_NAME")
+            or libretto["api_name"]  # legacy
+        )
+        libretto["api_key"] = (
+            libretto["api_key"] or self.config.api_key or os.environ.get("LIBRETTO_API_KEY")
+        )
+        libretto["chat_id"] = (
+            libretto["chat_id"] or self.config.chat_id or os.environ.get("LIBRETTO_CHAT_ID")
+        )
+        libretto["project_key"] = libretto["project_key"] or os.environ.get("LIBRETTO_PROJECT_KEY")
+        libretto["feedback_key"] = libretto["feedback_key"] or str(uuid.uuid4())
+
+        return libretto
+
+    def _build_model_params(self, **original_kwargs) -> Dict[str, Any]:
+        model_params = {"modelProvider": "openai"}
+        for k, v in original_kwargs.items():
+            if v == NOT_GIVEN or v is None:
```

Comment:
minor but should this be `v is NOT_GIVEN`? I'm assuming this is some well known sentinel and this might avoid odd coersion?